### PR TITLE
Measure compressed key size & benchmark decompression

### PR DESCRIPTION
**Pull request by soonum**

### CODE REVIEW

File: tfhe/benches/shortint/bench.rs [hunk 1]
```diff
@@ -304,6 +306,39 @@ fn programmable_bootstrapping(c: &mut Criterion) {
     bench_group.finish();
 }
 
+fn server_key_from_compressed_key(c: &mut Criterion) {
+    let mut bench_group = c.benchmark_group("uncompress_key");
+    bench_group
+        .sample_size(10)
+        .measurement_time(std::time::Duration::from_secs(60));
+
+    for param in SERVER_KEY_BENCH_PARAMS {
```
Comment (IceTDrinker) [inline tfhe/benches/shortint/bench.rs:315]: no multi bit ? don't remember how we manage benchmarks params

File: tfhe/benches/shortint/bench.rs [hunk 2]
```diff
@@ -304,6 +306,39 @@ fn programmable_bootstrapping(c: &mut Criterion) {
...
+        let param: PBSParameters = param.into();
+        let keys = KEY_CACHE.get_from_param(param);
+        let sks_compressed = CompressedServerKey::new(&keys.client_key());
+
+        let bench_id = format!("shortint::uncompress_key::{}", param.name());
+
+        bench_group.bench_function(&bench_id, |b| {
+            b.iter(|| {
+                let _ = ServerKey::from(sks_compressed.clone());
```
Comment (IceTDrinker) [inline tfhe/benches/shortint/bench.rs:324]: the clone is going to give wrong results as the clone operation is going to appear in the global time of the benchmark
we need an iter with input preparation I believe, integer benchmarks have it IIRC or maybe it's also in that file

**Review by IceTDrinker — CHANGES_REQUESTED** [review]
The benchmarks timings will be wrong because of the clone in the iter closure, we need to use the same patterns as for integer benchmarks where the input is prepared beforehand

File: tfhe/benches/shortint/bench.rs (hunk 1 above)
Comment (soonum) [reply tfhe/benches/shortint/bench.rs:315]: Thing is we don't have any multi-bit parameters in shortint. Should I move this benchmark to integer layer then ?

Comment (IceTDrinker) [reply tfhe/benches/shortint/bench.rs:315]: There should be multi bits params in shortint, check parameters::multi_bit or similar

File: tfhe/benches/shortint/bench.rs (hunk 2 above)
Comment (IceTDrinker) [reply tfhe/benches/shortint/bench.rs:324]: perfect

**Comment by soonum** [comment]
@slab-ci shortint_bench

File: .github/workflows/shortint_benchmark.yml [hunk 3]
```diff
@@ -77,6 +77,20 @@ jobs:
           --name-suffix avx512 \
           --throughput
 
+      - name: Run multi-bit benchmarks with AVX512
+        run: |
+          rm -rf target/criterion
+          make AVX512_SUPPORT=ON bench_shortint_multi_bit
```
Comment (IceTDrinker) [inline .github/workflows/shortint_benchmark.yml:83]: I don't know if this is the right PR for this, as there were some slow downs/some runs took longer as we did not have enough cores

File: tfhe/benches/shortint/bench.rs [hunk 4]
```diff
@@ -304,6 +340,47 @@ fn programmable_bootstrapping(c: &mut Criterion) {
...
+        let keys = KEY_CACHE.get_from_param(*param);
+        let sks_compressed = CompressedServerKey::new(&keys.client_key());
+
+        let bench_id = format!("shortint::uncompress_key::{}", param.name());
+
+        bench_group.bench_function(&bench_id, |b| {
+            let clone_compressed_key = || {
+                let sks_cloned = sks_compressed.clone();
+                sks_cloned
+            };
```
Comment (IceTDrinker) [inline tfhe/benches/shortint/bench.rs:359]: let clone_compressed_key = || sks_compressed.clone(); should be enough I think (clippy will likely yell at you)

File: .github/workflows/shortint_benchmark.yml (hunk 3 above)
Comment (soonum) [reply .github/workflows/shortint_benchmark.yml:83]: What do you mean? I've made several clear commits in this PR , you want me to move the last commit to a new PR ?

**Review by IceTDrinker — COMMENTED** [review]
Thanks, my suggestion for the multi bit was mainly for the decompression of the keys, I'm not sure we want to bench all ops with all sizes for multi bit ? unless it was ready but I don't remember to be honest

File: .github/workflows/shortint_benchmark.yml (hunk 3 above)
Comment (IceTDrinker) [reply .github/workflows/shortint_benchmark.yml:83]: I think so, adding the decompression is easy, while there may be issues with the multi bit benchmarks ? Ah but you added them only to shortint, well if the timings are not crazy on shortint then should be fine !

File: tfhe/benches/shortint/bench.rs [hunk 5]
```diff
@@ -333,309 +410,333 @@ fn _bench_wopbs_param_message_8_norm2_5(c: &mut Criterion) {
...
+define_custom_bench_fn!(function_name: carry_extract, BenchParamsSet::Standard);
+
+define_custom_bench_fn!(
+    function_name: programmable_bootstrapping,
+    BenchParamsSet::Standard
+);
+
+define_custom_bench_fn!(
+    function_name: server_key_from_compressed_key,
+    BenchParamsSet::Standard
```
Comment (IceTDrinker) [inline tfhe/benches/shortint/bench.rs:739]: do we want the extended set for this maybe ?

File: tfhe/benches/shortint/bench.rs (hunk 4 above)
Comment (soonum) [reply tfhe/benches/shortint/bench.rs:359]: Ok but how I get my hand on the value? `clone_compressed_key` is a closure not the value returned from the cloning.

File: tfhe/benches/shortint/bench.rs (hunk 5 above)
Comment (soonum) [reply tfhe/benches/shortint/bench.rs:739]: depends on how long the bench would be.

File: tfhe/benches/shortint/bench.rs (hunk 4 above)
Comment (IceTDrinker) [reply tfhe/benches/shortint/bench.rs:359]: clone_compressed_key()

one liner closures have the right to omit curly braces

**Comment by soonum** [comment]
> Thanks, my suggestion for the multi bit was mainly for the decompression of the keys, I'm not sure we want to bench all ops with all sizes for multi bit ? unless it was ready but I don't remember to be honest

Maybe we could do it only for uncompress server key bench and leave the other.

**Comment by soonum** [comment]
@slab-ci shortint_bench

**Comment by soonum** [comment]
@slab-ci shortint_bench

File: tfhe/benches/shortint/bench.rs (hunk 5 above)
Comment (soonum) [reply tfhe/benches/shortint/bench.rs]: Extended in both classic and multi-bit PBS is now implemented.

File: .github/workflows/shortint_benchmark.yml (hunk 3 above)
Comment (soonum) [reply .github/workflows/shortint_benchmark.yml]: I remove the multi-bit shortint in the benchmark for now.

**Review by IceTDrinker — APPROVED** [review]
Thanks a lot !

**Pull request merged**